File: mps-n/src/mps-clock/StopWatchPS.cpp
/*
 * Copyright (C) 2015 by Laurentiu-Gheorghe Crisan
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * As a special exception, you may use this file as part of a free
 * software library without restriction. Specifically, if other files
 * instantiate templates or use macros or inline functions from this
 * file, or you compile this file and link it with other files to
 * produce an executable, this file does not by itself cause the
 * resulting executable to be covered by the GNU General Public
 * License. This exception does not however invalidate any other
 * reasons why the executable file might be covered by the GNU Library
 * General Public License.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "StopWatchPS.h"
#include "StopWatchSignal.h"
#include <mps/core/TriggerPort.h>
#include <Pt/System/Clock.h>
#include <mps/core/SignalList.h>

namespace mps{
namespace clock{

StopWatchPS::StopWatchPS(void)
: _counting(false)
{
}

StopWatchPS::~StopWatchPS(void)
{
}

void StopWatchPS::onInitialize()
{
    SynchSourcePS::onInitialize();

    _triggerPort = (mps::core::TriggerPort*) _inputPorts->at(0);
    
    if(_inputPorts->size() > 1)
    {
        mps::core::TriggerPort* resetPort = (mps::core::TriggerPort*) _inputPorts->at(1);

        resetPort->onEventTrigger += Pt::slot(*this, &StopWatchPS::onResetTrigger);
    }

    _triggerPort->onStartTrigger += Pt::slot( *this, &StopWatchPS::onStartTrigger);
    _triggerPort->onStopTrigger += Pt::slot( *this, &StopWatchPS::onStopTrigger);
}

void StopWatchPS::onStart()
{
    _counting = false;
    _deltaTime.setNull();

    if(_triggerPort->triggerType() ==  mps::core::TriggerPort::StopTrigger ||
        _triggerPort->triggerType() == mps::core::TriggerPort::NoTrigger)
        onStartTrigger();

    _triggerPort->start();
    SynchSourcePS::onStart();
}

void StopWatchPS::onSourceEvent(Pt::uint32_t& noOfRecords, Pt::uint32_t maxNoOfSamples, Pt::uint32_t sourceIdx, Pt::uint32_t portIdx, Pt::uint8_t* data)
{
    double time = 0;

    if(_counting)
        time = getTimeSpan(Pt::System::Clock::getSystemTicks() - _startTime);
    else
        time = getTimeSpan(_deltaTime);

    for( Pt::uint32_t i = 0; i < noOfRecords; ++i)
        memcpy(&data[i*sizeof(double)],&time, sizeof(double));
}

double StopWatchPS::getTimeSpan(Pt::Timespan span)
{
    const mps::core::Port* port = _outputPorts->at(0);
    const StopWatchSignal* signal = (const StopWatchSignal*)port->signalList()->at(0);

    double time = 0.0;

    switch(signal->unitIdx())
    {
        case 0: //Milliseconds
            time = (double) span.toUSecs()/1000.0;
        break;
        
        case 1: //Second
            time  = (double) span.toUSecs() / 1000000.0;
        break;
        
        case 2: // Minute
            time = ((double) span.toUSecs() / 1000000.0)/60.0;
        break;

        case 3: //Hour
            time = (((double) span.toUSecs() / 1000000.0)/60.0)/60.0;
        break;
    }

    return time;
}

void StopWatchPS::onStartTrigger()
{
    _counting = true;
    _startTime = Pt::System::Clock::getSystemTicks();
    _deltaTime.setNull();
    _clock.start();
}

void StopWatchPS::onResetTrigger()
{
    _currentTime.setNull();
    _deltaTime.setNull();
    _clock.start();	
    _startTime = Pt::System::Clock::getSystemTicks();
}

void StopWatchPS::onStopTrigger()
{
    if(_counting)
    {
        _deltaTime = _clock.stop();
        _currentTime = Pt::System::Clock::getSystemTicks() - _startTime;
        _counting = false;
    }
}


void StopWatchPS::onUpdateDataValue( Pt::uint32_t noOfRecords, Pt::uint32_t sourceIdx, const mps::core::Port* port, const Pt::uint8_t* data )
{

}

}}

File: mps-m/Mp.Mod.Clock/StopWatchPS.cs
//    MeaProcess - Meaurement and Automation framework.
//    Copyright (C) 2015  Laurentiu-Gheorghe Crisan
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.

using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using Mp.Scheme.Sdk;
using Mp.Utils;

namespace Mp.Mod.Clock
{
    public class StopWatchPS : ProcessStation
    {
        public StopWatchPS()
        {
            base.Type = "Mp.Clock.PS.StopWatch";
            base.Text = StringResource.StopWatch;
            base.Group = StringResource.Time;
            base.Symbol = Images.Timer;
            base.Icon = Images.TimerIcon;
        }

        public override string RuntimeModule
        {
            get { return "mps-clock"; }
        }

        public override void OnLoadResources()
        {
            base.Text = StringResource.StopWatch;
            base.Group = StringResource.Time;
        }

        public override string Description
        {
            get
            {
                return StringResource.StopWatchPsDescription;
            }
        }

        public override void OnDefaultInit()
        {
            base.OnDefaultInit();

            //Create the data out port.
            Port port = new Port(new Point(_rectangle.Right + PortWidth, (int)(_rectangle.Top + PortTopOffset)), "Mp.Port.Out", false, true);
            port.SignalList = Document.CreateSignalList();
            InitMenuForPort(port);
            AddPort(port);

            //Create the trigger in port
            port = new Port(new Point(_rectangle.Left - PortWidth, (int)(_rectangle.Top + PortTopOffset)), "Mp.Port.Trigger", true, false);
            port.ConnectorBrush = new SolidBrush(Color.DarkGreen);
            InitMenuForPort(port);
            AddPort(port);

            //Create the reset in port
            port = new Port(new Point(_rectangle.Left - PortWidth, (int)(_rectangle.Top + PortTopOffset + DistanceBetweenPort)), "Mp.Port.Trigger", true, false);
            port.ConnectorBrush = new SolidBrush(Color.DarkGreen);
            AddPort(port);

            XmlHelper.SetParamNumber(port.XmlRep, "triggerType", "uint8_t", 4);
        }

        private void InitMenuForPort(Port port)
        {
            //Create the context menu.
            port.ContextMenuStrip = new ContextMenuStrip();
            port.ContextMenuStrip.Tag = port;

            ToolStripMenuItem menuItem = new ToolStripMenuItem(StringResource.MenuProperties);

            menuItem.Font = new System.Drawing.Font("Tahoma", 8.25F, System.Drawing.FontStyle.Bold);
            menuItem.Click += new System.EventHandler(this.OnPropertyDataPort);
            port.ContextMenuStrip.Items.Add(menuItem);
        }

        public override void OnLoadXml()
        {
            base.OnLoadXml();

            InitMenuForPort(InputPorts[0]);
            InitMenuForPort(OutputPorts[0]);
        }

        public override void OnPortDoubleClick(Port port)
        {
            base.OnPortDoubleClick(port);                        
            ShowPortDlg(port);
        }

        protected void OnPropertyDataPort(object sender, EventArgs e)
        {
            ToolStripMenuItem item = (sender as ToolStripMenuItem);

            Port port = (item.Owner.Tag as Port);

            ShowPortDlg(port);     
        }

        private void ShowPortDlg(Port port)
        {
            if (port.Type == "Mp.Port.Out")
            {
                StopWatchOutPortDlg dlg = new StopWatchOutPortDlg(Document, port.SignalList);
                dlg.ShowDialog();
            }
            else
            {
                if (InputPorts[0] == port)
                {
                    SimpleTriggerPortDlg dlg = new SimpleTriggerPortDlg(Document, port.XmlRep);
                    dlg.ShowDialog();
                }
            }
        }

        protected override void OnValidate(List<ValidationInfo> valInfoList)
        {
            base.OnValidate(valInfoList);

            Port dataOutPort = OutputPorts[0];

            if (!dataOutPort.Connected)
            {
                string msg = String.Format(StringResource.DataOutPortConError, this.Text);
                valInfoList.Add(new ValidationInfo(msg, ValidationInfo.InfoType.Error));
            }

            if (dataOutPort.SignalList.InnerText == "")
            {
                string msg = String.Format(StringResource.DataOutPortSigError, this.Text);
                valInfoList.Add(new ValidationInfo(msg, ValidationInfo.InfoType.Error));
            }

            Port triggerPort = InputPorts[0];

            int triggerType = (int)XmlHelper.GetParamNumber(triggerPort.XmlRep, "triggerType");

            if (triggerType == 0)
                return;

            if (!triggerPort.Connected)
            {
                string msg = String.Format(StringResource.TriggerPortNotConErr, this.Text);
                valInfoList.Add(new ValidationInfo(msg, ValidationInfo.InfoType.Error));
                return;
            }

            switch (triggerType)
            {
                case 1:
                case 2:
                    if (triggerPort.SignalList.ChildNodes.Count < 1)
                    {
                        string msg = String.Format(StringResource.TriggerOneSigErr, this.Text);
                        valInfoList.Add(new ValidationInfo(msg, ValidationInfo.InfoType.Error));
                    }
                break;

                case 3:
                {
                    bool oneStartStop = XmlHelper.GetParamNumber(triggerPort.XmlRep, "oneStartStopSignal") > 0;

                    if (triggerPort.SignalList.ChildNodes.Count < 2 && !oneStartStop)
                    {
                        string msg = String.Format(StringResource.TriggerTwoSigErr,this.Text);
                        valInfoList.Add(new ValidationInfo(msg, ValidationInfo.InfoType.Error));
                    }

                    if (triggerPort.SignalList.ChildNodes.Count < 1 && oneStartStop)
                    {
                        string msg = String.Format(StringResource.TriggerOneSigErr, this.Text);
                        valInfoList.Add(new ValidationInfo(msg, ValidationInfo.InfoType.Error));
                    }
                }
                break;
            }
        }

        public override void OnHelpRequested()
        {
            Document.ShowHelp(this.Site, 570);
        }
    }
}
